File: tools/Sources/CommitKit/CommitConfig.swift
/// Conventional-commit vocabulary. Defaults from https://www.conventionalcommits.org/.
public enum CommitDefaults {
    public static let types: [String] = [
        "feat",
        "fix",
        "build",
        "chore",
        "ci",
        "docs",
        "style",
        "refactor",
        "perf",
        "test",
        "revert",
    ]

    public static let maxLength = 72
}

public struct CommitConfig: Equatable, Sendable {
    public var types: [String]
    /// Categorised allowed scopes; `nil` means any scope is permitted.
    public var scopes: [String: [String]]?
    public var maxLength: Int

    public init(types: [String], scopes: [String: [String]]? = nil, maxLength: Int) {
        self.types = types
        self.scopes = scopes
        self.maxLength = maxLength
    }

    public static let `default` = CommitConfig(
        types: CommitDefaults.types,
        maxLength: CommitDefaults.maxLength)

    /// Flatten every configured scope category, or `nil` when no scopes are set.
    public var allScopes: [String]? {
        guard let scopes else { return nil }
        return scopes.values.flatMap { $0 }
    }

    /// The active configuration. Kept as a function to mirror `loadConfig()` and
    /// leave room for reading project config later.
    public static func load() -> CommitConfig { .default }
}

File: tools/commit/commit.config.ts
// Conventional-commit vocabulary. Defaults from https://www.conventionalcommits.org/.
export const DEFAULT_TYPES: string[] = [
  'feat',
  'fix',
  'build',
  'chore',
  'ci',
  'docs',
  'style',
  'refactor',
  'perf',
  'test',
  'revert',
]

export const DEFAULT_MAX_LENGTH: number = 72

export type CommitConfig = {
  types: string[]
  scopes?: Record<string, string[]>
  maxLength: number
}

export const defaultConfig: CommitConfig = {
  types: DEFAULT_TYPES,
  maxLength: DEFAULT_MAX_LENGTH,
}

export const getAllScopes = (config: CommitConfig): string[] | undefined => {
  if (!config.scopes) {
    return undefined
  }

  return Object.values(config.scopes).flat()
}

export const loadConfig = (): CommitConfig => {
  return defaultConfig
}
